File: Server/src/main/java/core/game/interaction/inter/SilverInterface.java
package core.game.interaction.inter;

import core.cache.def.impl.ItemDefinition;
import core.game.component.Component;
import core.game.component.ComponentDefinition;
import core.game.component.ComponentPlugin;
import core.game.node.entity.player.link.diary.DiaryType;
import core.game.world.map.Location;
import core.plugin.Initializable;
import core.game.node.entity.skill.Skills;
import core.game.node.entity.skill.crafting.SilverProduct;
import core.game.node.entity.player.Player;
import core.game.node.entity.player.link.RunScript;
import core.game.node.item.Item;
import core.game.system.task.Pulse;
import core.game.world.GameWorld;
import core.game.world.update.flag.context.Animation;
import core.plugin.Plugin;

/**
 * Represents the silver crafting making interface.
 * @author 'Vexia
 */
@Initializable
public final class SilverInterface extends ComponentPlugin {

	/**
	 * Represents the silver bar item.
	 */
	private static final Item SILVER_BAR = new Item(2355);

	@Override
	public Plugin<Object> newInstance(Object arg) throws Throwable {
		ComponentDefinition.put(438, this);
		return this;
	}

	@Override
	public boolean handle(final Player p, Component component, int opcode, int button, int slot, int itemId) {
		final SilverProduct silver = SilverProduct.forId(button);
		if (silver == null) {
			return true;
		}
		if (!p.getInventory().contains(silver.getNeeded(), 1)) {
			p.getPacketDispatch().sendMessage("You need a " + ItemDefinition.forId(silver.getNeeded()).getName().toLowerCase() + " to make this item.");
			return true;
		}
		if (silver == SilverProduct.SILVTHRIL_ROD) {
			p.getPacketDispatch().sendMessage("You can't do that yet.");
			return true;
		}
		if (p.getSkills().getLevel(Skills.CRAFTING) < silver.getLevel()) {
			p.getPacketDispatch().sendMessage("You need a crafting level of " + silver.getLevel() + " to make this.");
		}
		int amt = -1;
		switch (opcode) {
		case 155:
			amt = 1;
			break;
		case 196:
			amt = 5;
			break;
		case 124:
			amt = p.getInventory().getAmount(new Item(2355));
			break;
		case 199:
			p.setAttribute("runscript", new RunScript() {
				@Override
				public boolean handle() {
					int ammount = (int) value;
					make(p, silver, ammount);
					return true;
				}
			});
			p.getDialogueInterpreter().sendInput(false, "Enter the amount.");
			return true;
		}
		make(p, silver, amt);
		return true;
	}

	private void make(final Player player, final SilverProduct def, final int ammount) {
		if (!player.getInventory().containsItem(SILVER_BAR)) {
			return;
		}
		player.getInterfaceManager().close();
		GameWorld.getPulser().submit(new Pulse(1, player) {
			int amt = ammount;

			@Override
			public boolean pulse() {
				if (amt < 1) {
					return true;
				}
				if (!player.getInventory().contains(def.getNeeded(), 1)) {
					return true;
				}
				if (player.getInventory().containsItem(SILVER_BAR) && player.getInventory().contains(def.getNeeded(), 1)) {
					player.animate(new Animation(899));
					if (player.getInventory().remove(SILVER_BAR)) {
						player.getInventory().add(new Item(def.getProduct(), def == SilverProduct.SILVER_BOLTS ? 10 : 1));
						player.getSkills().addExperience(Skills.CRAFTING, def.getExp(), true);
					}

					// Craft a holy symbol in the Lumbridge furnace
					if (def == SilverProduct.UNBLESSED && player.getLocation().withinDistance(Location.create(3226, 3254, 0))) {
						player.getAchievementDiaryManager().finishTask(player, DiaryType.LUMBRIDGE, 2, 8);
					}
				} else {
					return true;
				}
				amt--;
				return false;
			}

			@Override
			public void stop() {
				player.animate(new Animation(-1));
			}
		});
	}
}

File: Server/src/main/kotlin/rs09/game/interaction/inter/SilverInterface.kt
package rs09.game.interaction.inter

import api.*
import core.game.node.entity.player.Player
import core.game.node.entity.player.link.diary.DiaryType
import core.game.node.entity.skill.Skills
import core.game.node.entity.skill.crafting.SilverProduct
import core.game.system.task.Pulse
import org.rs09.consts.Components
import org.rs09.consts.Items
import rs09.game.interaction.InteractionListener
import rs09.game.interaction.InterfaceListener

class SilverInterface : InterfaceListener() {

    val IFACE = Components.CRAFTING_SILVER_CASTING_438
    val ANIM = getAnimation(899)

    override fun defineListeners() {
        onOpen(IFACE){player, _ ->
            sendItemOnInterface(player, IFACE, 17, 1718)
            sendItemOnInterface(player, IFACE, 24, 1724)
            sendItemOnInterface(player, IFACE, 31, 2961)
            sendItemOnInterface(player, IFACE, 38, 4201)
            sendItemOnInterface(player, IFACE, 45, 5525)
            sendItemOnInterface(player, IFACE, 53, 7637)
            sendItemOnInterface(player, IFACE, 60, 6748)
            sendItemOnInterface(player, IFACE, 67, 9382)
            sendItemOnInterface(player, IFACE, 74, 13154)
            return@onOpen true
        }

        on(IFACE){player, _, opcode, buttonID, _, _ ->
            val product = SilverProduct.forId(buttonID) ?: return@on true

            if(!inInventory(player, product.needed)){
                sendMessage(player, "You need a ${itemDefinition(product.needed).name.toLowerCase()} to make this item.")
                return@on true
            }

            if(product == SilverProduct.SILVTHRIL_ROD || product == SilverProduct.SILVTHRIL_CHAIN){
                sendMessage(player, "You can't do that yet.")
                return@on true
            }

            if(!hasLevelDyn(player, Skills.CRAFTING, product.level)){
                sendMessage(player, "You need a Crafting level of ${product.level} to make this.")
                return@on true
            }

            val amt = when(opcode){
                155 -> 1
                196 -> 5
                124 -> amountInInventory(player, Items.SILVER_BAR_2355)
                199 -> -1 //Make X
                else -> return@on true
            }

            if(amt == -1) sendInputDialogue(player, true, "Enter the amount:"){value ->
                make(player,product,value as Int)
            } else make(player, product, amt)
            return@on true
        }
    }

    fun make(player: Player, product: SilverProduct, amount: Int){
        var amt = amount
        closeInterface(player)
        submitIndividualPulse(player, object : Pulse(){
            override fun pulse(): Boolean {
                if(amt < 1) return true
                if(!inInventory(player, product.needed) || !inInventory(player, Items.SILVER_BAR_2355)) return true;
                animate(player, ANIM)

                if(removeItem(player, Items.SILVER_BAR_2355, api.Container.INVENTORY)){
                    addItem(player, product.product, if(product == SilverProduct.SILVER_BOLTS) 10 else 1)
                    rewardXP(player, Skills.CRAFTING, product.exp)

                    if(product == SilverProduct.UNBLESSED && player.location.withinDistance(location(3226,3254,0))){
                        player.achievementDiaryManager.finishTask(player, DiaryType.LUMBRIDGE, 2, 8)
                    }
                    delay = 3
                } else return true

                amt--
                return false
            }
        })
    }
}

class SilverBarUseWith : InteractionListener() {
    val FURNACES = intArrayOf(2966, 3044, 3294, 4304, 6189, 11009, 11010, 11666, 12100, 12809, 18497, 18525, 18526, 21879, 22721, 26814, 28433, 28434, 30021, 30510, 36956, 37651)
    override fun defineListeners() {
        onUseWith(SCENERY, Items.SILVER_BAR_2355, *FURNACES){ player, _, _ ->
            openInterface(player, Components.CRAFTING_SILVER_CASTING_438)
            return@onUseWith true
        }
    }
}